### PR TITLE
[moe]: fix ep/tp tests, add hierarchical all2all

### PR DESCRIPTION
**Pull request by CWHer**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/tensor/moe_tensor/moe_info.py [hunk 1]
```diff
@@ -15,12 +15,8 @@ def __init__(self, ep_inside: bool, ep_size: int, dp_size: int, pp_size: int = 1
...
         """
         self.pp_size, self.dp_size, self.ep_size = pp_size, dp_size, ep_size
-        if ep_inside:
-            self.pp_axis, self.dp_axis, self.ep_axis = 0, 1, 2
-            self.pg = ProcessGroupMesh(self.pp_size, self.dp_size, self.ep_size)
-        else:
-            self.pp_axis, self.ep_axis, self.dp_axis = 0, 1, 2
-            self.pg = ProcessGroupMesh(self.pp_size, self.ep_size, self.dp_size)
+        self.pp_axis, self.dp_axis, self.ep_axis = (0, 1, 2) if ep_inside else (0, 2, 1)
+        self.pg = ProcessGroupMesh(self.pp_size, self.dp_size, self.ep_size)
```
Comment (oahzxl) [inline colossalai/tensor/moe_tensor/moe_info.py:19]: Process group should support dp inside or ep size inside. You can add an assert if hierachical alltoall does not support one of them

File: tests/test_moe/test_moe_ep_tp.py [hunk 2]
```diff
@@ -21,7 +132,12 @@ def run_test(rank: int, world_size: int, port: int, num_experts: int, batch_size
     local_model = SparseMLP(num_experts=num_experts, hidden_size=dim, intermediate_size=dim * 2)
     MOE_MANAGER.__init__()
     MOE_MANAGER.setup(seed, parallel="EP")
-    ep_model = SparseMLP(num_experts=num_experts, hidden_size=dim, intermediate_size=dim * 2)
+    ep_model = SparseMLP(
+        num_experts=num_experts,
+        hidden_size=dim,
+        intermediate_size=dim * 2,
+        create_hierarchical_group=lambda group: create_ep_hierarchical_group(group, num_node=1, nproc_per_node=2)
```
Comment (oahzxl) [inline tests/test_moe/test_moe_ep_tp.py:139]: Can we auto set this?

File: colossalai/moe/_operation.py [hunk 3]
```diff
@@ -138,8 +138,66 @@ def forward(
...
+
+        if dist.get_rank() == src_rank:
+            # intra-node gather
+            intra_output = [torch.empty_like(inputs) for _ in range(local_world_size)]
+            dist.gather(inputs, intra_output, dst=src_rank, group=intra_node_group)
+
+            intra_output = [v.chunk(world_size, dim=0) for v in intra_output]
+            intra_output = torch.cat(sum(zip(*intra_output), ()))
+
+            # inter-node all-to-all
```
Comment (oahzxl) [inline colossalai/moe/_operation.py:177]: After inter node allgather, can we squeeze its empty space before alltoall? Then restore its layout when alltoall send back.

File: colossalai/moe/layers.py [hunk 4]
```diff
@@ -104,6 +105,7 @@ def __init__(
         if self.expert_parallel is not None:
             self.ep_group = get_ep_group(self.experts)
             self.ep_size = get_ep_size(self.experts)
+            self.ep_hierarchical_group = create_hierarchical_group(self.ep_group)
```
Comment (oahzxl) [inline colossalai/moe/layers.py:108]: Use an arg to use hierachical alltoall. Also need to add it in the config file and set_openmoe_arg function.

File: tests/test_moe/test_moe_ep_tp.py (hunk 2 above)
Comment (CWHer) [reply tests/test_moe/test_moe_ep_tp.py]: This can be auto-set if `torchrun` is used.

> `LOCAL_WORLD_SIZE` - The local world size (e.g. number of workers running locally); equals to `--nproc-per-node` specified on `torchrun`. 

9795011 resolves this.

File: colossalai/moe/layers.py (hunk 4 above)
Comment (CWHer) [reply colossalai/moe/layers.py]: c6d51c0 resolves this.

File: examples/language/openmoe/benchmark/benchmark_cai.py [hunk 5]
```diff
@@ -213,7 +215,8 @@ def main():
         moe_layer_interval=config.moe_layer_interval,
         enable_load_balance=args.load_balance,
         enable_kernel=args.use_kernel,
-        enable_comm_overlap=args.overlap_alltoall,
+        enable_comm_overlap=args.overlap_comm,
+        enable_hierarchical_alltoall=args.hierarchical_alltoall,
```
Comment (oahzxl) [inline examples/language/openmoe/benchmark/benchmark_cai.py:219]: you should also add arg in this function

File: colossalai/tensor/moe_tensor/moe_info.py (hunk 1 above)
Comment (CWHer) [reply colossalai/tensor/moe_tensor/moe_info.py:19]: Hierarchical all-to-all communication can support both of the mentioned scenarios. As long as the experts are sharded by their rank in ascending order. For example, if `ep_group = [0, 1, 8, 9]`, then rank 0 will contain expert shard 0, rank 1 will contain expert shard 1, and so on.

Comment (oahzxl) [reply colossalai/tensor/moe_tensor/moe_info.py:19]: You fix the processgroupmesh order in `self.pg = ProcessGroupMesh(self.pp_size, self.dp_size, self.ep_size)`. It should change with dp and ep axis

File: colossalai/moe/_operation.py (hunk 3 above)
Comment (oahzxl) [reply colossalai/moe/_operation.py:177]: You can also can add a todo here and implement in next pr

File: colossalai/tensor/moe_tensor/moe_info.py (hunk 1 above)
Comment (CWHer) [reply colossalai/tensor/moe_tensor/moe_info.py]: I changed it unintentionally. d93b7709 revert this.

File: colossalai/moe/_operation.py (hunk 3 above)
Comment (CWHer) [reply colossalai/moe/_operation.py:178]: 38b7bbd marks it as TODO for now.

**Review by ver217 — APPROVED** [review]

File: tests/test_moe/test_moe_ep_tp.py [hunk 6]
```diff
@@ -21,7 +133,14 @@ def run_test(rank: int, world_size: int, port: int, num_experts: int, batch_size
     local_model = SparseMLP(num_experts=num_experts, hidden_size=dim, intermediate_size=dim * 2)
     MOE_MANAGER.__init__()
     MOE_MANAGER.setup(parallel="EP")
-    ep_model = SparseMLP(num_experts=num_experts, hidden_size=dim, intermediate_size=dim * 2)
+    os.environ["LOCAL_WORLD_SIZE"] = str(world_size)
+    enable_hierarchical_comm = torch.__version__ >= "1.13.1"
```
Comment (oahzxl) [inline tests/test_moe/test_moe_ep_tp.py:137]: You can use processgroupmesh to get group ranks instead of restricting torch version

File: tests/test_moe/test_moe_ep_tp.py [hunk 7]
```diff
@@ -21,7 +133,14 @@ def run_test(rank: int, world_size: int, port: int, num_experts: int, batch_size
...
+    ep_model = SparseMLP(
+        num_experts=num_experts,
+        hidden_size=dim,
+        intermediate_size=dim * 2,
+        enable_hierarchical_comm=enable_hierarchical_comm
```
Comment (oahzxl) [inline tests/test_moe/test_moe_ep_tp.py:142]: all of two all2all methods should be tested

File: tests/test_moe/test_moe_ep_tp.py [hunk 8]
```diff
@@ -34,48 +153,76 @@ def run_test(rank: int, world_size: int, port: int, num_experts: int, batch_size
...
 
 @pytest.mark.dist
-@pytest.mark.parametrize("num_experts", [4, 8])
-@pytest.mark.parametrize("batch_size", [4])
-@pytest.mark.parametrize("dim", [32])
-@pytest.mark.parametrize("seed", [42])
+@pytest.mark.parametrize("num_experts", [4, 64])
+@pytest.mark.parametrize("batch_size", [16])
+@pytest.mark.parametrize("dim", [64])
+@pytest.mark.parametrize("seed", [42, 127])
```
Comment (oahzxl) [inline tests/test_moe/test_moe_ep_tp.py:221]: we do not use seed now
